### PR TITLE
Added user realization together with pass hashing

### PR DESCRIPTION
**Pull request by RLudchenko**

### CODE REVIEW

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            session.save(user);
+            transaction.commit();
+            LOGGER.info(session + " was added to DB");
+            return user;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("There was an error adding "
+                    + session, e);
```
Comment (boroda4436) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: чому сессію прокидуєш в помилку? В чому користь?

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            CriteriaBuilder cb = session.getCriteriaBuilder();
+            CriteriaQuery<User> criteriaQuery = cb.createQuery(User.class);
+            Root<User> root = criteriaQuery.from(User.class);
+            Predicate predicate = cb
+                    .equal(root.get("email"), email);
```
Comment (boroda4436) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: please inline

File: src/main/java/com/dev/cinema/util/HashUtil.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            MessageDigest messageDigest = MessageDigest.getInstance(hasher);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+
+            for (byte b : digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException();
```
Comment (boroda4436) [inline src/main/java/com/dev/cinema/util/HashUtil.java]:
```suggestion
            throw new RuntimeException(ex);
```

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java (hunk 1 above)
Comment (RLudchenko) [reply src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: Fixed!

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java (hunk 2 above)
Comment (RLudchenko) [reply src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: Done

File: src/main/java/com/dev/cinema/util/HashUtil.java (hunk 3 above)
Comment (RLudchenko) [reply src/main/java/com/dev/cinema/util/HashUtil.java]: Done

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import org.hibernate.Transaction;
+
+@Dao
+public class UserDaoImpl implements UserDao {
+    private static final Logger LOGGER = LogManager.getLogger(UserDaoImpl.class);
+
+    @Override
+    public User add(User user) {
+        Session session = null;
+        Transaction transaction = null;
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java:25]: let's move declaration to try block

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            }
+            throw new DataProcessingException("There was an error adding ", e);
+        } finally {
+            session.close();
+        }
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java:45]: agreed to use try with finally

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        User userFromDB = userService.findByEmail(email).orElseThrow(() ->
+                new AuthenticationException("Incorrect username or password!"));
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: let's make user service return user or null

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import com.dev.cinema.lib.Inject;
+import com.dev.cinema.lib.Service;
+import com.dev.cinema.model.User;
+import com.dev.cinema.service.UserService;
+import java.util.Optional;
+
+@Service
+public class UserServiceImpl implements UserService {
+    @Inject
+    UserDao userDao;
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: access modifier

File: src/main/resources/hibernate.cfg.xml [hunk 8]
```diff
@@ -15,5 +15,6 @@
         <mapping class="com.dev.cinema.model.Movie"/>
         <mapping class="com.dev.cinema.model.MovieSession"/>
         <mapping class="com.dev.cinema.model.CinemaHall"/>
+        <mapping class="com.dev.cinema.model.User"/>
     </session-factory>
 </hibernate-configuration>
```
Comment (dmytro-hryhoruk) [inline src/main/resources/hibernate.cfg.xml]: empty line

Comment (RLudchenko) [reply src/main/resources/hibernate.cfg.xml]: Done

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java (hunk 7 above)
Comment (RLudchenko) [reply src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: done

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java (hunk 6 above)
Comment (RLudchenko) [reply src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: Why does it need to return the user or null? In my opinion, it's better to leave an exception

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        Optional<User> getUserFromDB = userDao.findByEmail(email);
+        if (getUserFromDB.isPresent()) {
+            return getUserFromDB;
+        }
+
+        return null;
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: what the point of making check and return optional anyway?

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java (hunk 6 above)
Comment (dmytro-hryhoruk) [reply src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: our service should not return optional.So here we will check if(user != null && passwordsDoesn'tMatch) then throw new  Exception

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java (hunk 9 above)
Comment (RLudchenko) [reply src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: I accidentally did that, I changed it back

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -19,11 +19,6 @@ public User add(User user) {
 
     @Override
     public Optional<User> findByEmail(String email) {
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: service should not return optional)

Comment (RLudchenko) [reply src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: Done

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**